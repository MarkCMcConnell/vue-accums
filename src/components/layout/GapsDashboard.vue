<template>
  <section>
    <h3>Gaps in Care</h3>
    <DataTable :data="formatGaps" :fields="gapsFields"/>
  </section>
</template>

<script>
// General UI Components
import DataTable from '../ui/DataTable.vue'
// Mixins
import { dataHelpers } from '../../mixins/dataHelpers.js'

export default {
  components: {
    DataTable
  },
  props: {
    gapsData: {
      type: Array,
      required: true
    }
  },
  mixins: [dataHelpers],
  data () {
    return {
      gapsFields: [
        {
          label: 'Name',
          field: 'Name'
        },
        {
          label: 'Description',
          field: 'Accum_Desc'
        },
        {
          label: 'Recommended Frequency',
          field: 'RecFrequency'
        },
        {
          label: 'Status',
          field: 'ClaimSystemCode'
        },
        {
          label: 'Last Visit',
          field: 'Field1_Date'
        },
        {
          label: 'Covered Benefit',
          field: 'CovBenefit'
        }
      ]
    }
  },
  computed: {
    formatGaps () {
      const gaps = this.gapsData.filter(gap => {
        return gap.SpecificType === 'GC'
      })

      let dataValues = []
      
      for (let obj of gaps) {
        let data = {}
        // TODO: Handle null/undefined values more appropriately
        for (let value of this.gapsFields) {
          let currentKey = obj[value.field]
          // Capture dates and convert /Date()/ format to human readable
          if (currentKey.substring(0, 5) === '/Date') {
            currentKey = this.convertJSONDateToDateString(currentKey)
          }
          // Convert CovBenefit abbreviation to full word
          if (value.field === 'CovBenefit') {
            if (currentKey === 'Y') {
              currentKey = 'Yes'
            }
            if (currentKey === 'N') {
              currentKey === 'No'
            }
          }
          // Avoid setting an empty row
          if (currentKey !== '' && currentKey !== undefined) {
            data[value.field] = currentKey
          }
        }
        // Push to the array that will be passed to the coverages table
        dataValues.push(data)
      }
      return dataValues
    }
  }
}
</script>

<style>
</style>
