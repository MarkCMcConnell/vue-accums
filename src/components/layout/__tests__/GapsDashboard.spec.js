/* eslint-disable */
import { shallowMount } from '@vue/test-utils'

import GapsDashboard from '../GapsDashboard.vue'

describe('GapsDashboard', () => {
	let wrapper

	beforeEach(() => {
		wrapper = shallowMount(GapsDashboard, {
			propsData: {
				gapsData: [{
					Name: 'Diabetes Check',
					Accum_Desc: 'Blood Sugar Testing',
					RecFrequency: 'Weekly',
					ClaimSystemCode: 'Due Soon',
					Field1_Date: '\/Date(1528978573953-0400)\/',
					CovBenefit: 'Y',
					SpecificType: 'GC'
				}]
			}
		})
	})

	it('renders the component', () => {
		expect(wrapper.html()).toMatchSnapshot()
	})
	
	describe('formatGaps', () => {
		it('converts Y to Yes', () => {
			expect(wrapper.vm.formatGaps).toEqual(
				expect.arrayContaining([
					expect.objectContaining({
						CovBenefit: 'Yes'
					})
				])
			)
		})
	
		it('converts Healthx Date string to human readable', () => {
			expect(wrapper.vm.formatGaps).toEqual(
				expect.arrayContaining([
					expect.objectContaining({
						Field1_Date: '6/14/2018'
					})
				])
			)
		})
	})
})
