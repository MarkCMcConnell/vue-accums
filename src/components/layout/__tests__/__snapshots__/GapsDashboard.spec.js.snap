// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GapsDashboard renders the component 1`] = `
<section>
  <h3>Gaps in Care</h3>
  <datatable-stub fields="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" data="[object Object]"></datatable-stub>
</section>
`;
